File: code/java/剑指offer34.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    List<List<Integer>> res;

    public void dfs(TreeNode root,List<Integer> path,int value,int sum){
        path.add(root.val);
        value+=root.val;
        if(value==sum){
            if(root.left==null&&root.right==null) res.add(new ArrayList<>(path));
        }
        if(root.left!=null) dfs(root.left,path,value,sum);
        if(root.right!=null) dfs(root.right,path,value,sum);
        value-=root.val;
        path.remove(path.size()-1);
    }

    public List<List<Integer>> pathSum(TreeNode root, int sum) {
        if(root==null) return new ArrayList<>(0);
        res = new ArrayList<>();
        List<Integer> path = new ArrayList<>();
        dfs(root,path,0,sum);
        return res;
    }
}

File: code/python/剑指offer34.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def pathSum(self, root: TreeNode, sum: int) -> List[List[int]]:
        res = []
        def dfs(node,path,value):
            path.append(node.val)
            value+=node.val
            if value==sum:
                if node.left is None and node.right is None:
                    res.append(path.copy())
            if node.left is not None:
                dfs(node.left,path,value)
            if node.right is not None:
                dfs(node.right,path,value)
            path.pop()
            value-=node.val
        if root is None:
            return []
        dfs(root,[],0)
        return res